---
output:
  BiocStyle::html_document
---

<!-- rmarkdown v1 -->

<!--
%\VignetteIndexEntry{R Markdown Lecture}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
palette(RSkittleBrewer::RSkittleBrewer("tropical"))
```

# Exploratory Analysis

Package: [genstats](https://github.com/jtleek/genstats)<br />
Author: [Jeff Leek](www.jtleek.com) <br />
Compilation date: `r Sys.Date()`

### General principles

* Use plots as often as possible
* Use tables for phenotype data
* Look for 
  * Missing values
  * Outlier values
  * Mislabeled samples
  * Naming consistency

### Make the plots pretty

Install the [RSkittleBrewer](http://alyssafrazee.com/RSkittleBrewer.html) package if you haven't.  
```{r, eval=FALSE}
devtools::install_github("alyssafrazee/RSkittleBrewer")
```

Load the library and set up some values
```{r pretty}
library(RSkittleBrewer)
# Make the colors pretty
trop = RSkittleBrewer("tropical")
palette(trop)
```

### Load some data

We will use this expression set to look at how we use plots and tables to check for different characteristics

```{r}
library(Biobase)
con=url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata=pData(bm)
edata=exprs(bm)
fdata = fData(bm)
ls()
```

### Tables for factor/character variables

Tables are good for looking at factor or character variables, especially in phenotype data

```{r}
table(pdata$gender)
table(pdata$gender,pdata$race)
```

### Look for missing values

First check a summary of the distribution to look for scale, this is also one way to check for `NA` values.

```{r}
summary(edata)
```

`NA` is the most common character for missing values, but sometimes they are coded as spaces, 999, -1 or "missing". Check for missing values in a variety of ways

```{r}
# Use option useNA to include NA's in table
table(pdata$age,useNA="ifany")

# is.na checks for NA values
table(is.na(pdata$age))

# Check for other common missing names
sum(pdata$age==" ")

# Check genomic data for NAs
sum(is.na(edata))

# Make the distribution of NA's by genes
gene_na = rowSums(is.na(edata))
table(gene_na)

# Make the distribution of NA's by samples
sample_na = rowSums(is.na(edata))
table(sample_na)

```

### Make sure dimensions match up
The number of rows of the feature data should match the number of rows of the genomic data (both are the number of genes). The number of rows of the phenotype data should match the number of columns of the genomic data (both are the number of samples).
```{r}
dim(fdata)
dim(pdata)
dim(edata)
```

### Look at overall distributions

Here we see that there are a lot of outliers
```{r}
boxplot(log2(edata+1),col=2,range=0)
```

So we can remove rows that are mostly zero and notice any differences in the distributions across samples.

```{r}
library(dplyr)
edata = as.data.frame(edata)
filt_edata = filter(edata,rowMeans(edata) > 1)
boxplot(log2(filt_edata+1),col=2)
```

### Check for obvious data mixups

Here we are going to do a check to make sure that the men and women are correctly labeled by looking at expression on the Y chromosome. In general you might do several of this type of check to confirm the data are correctly labeled.  

Get the database package:

```{r,eval=FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("org.Hs.eg.db")
```


Get the chromosomes for each gene using the feature data. 
```{r}
library(org.Hs.eg.db)
aeid = as.character(fdata[,1])
chr = AnnotationDbi::select(org.Hs.eg.db,keys=aeid,keytype="ENSEMBL",columns="CHR")
head(chr)
```

Filter to the data on chromsome Y and sum up all the counts. A tricky issue is that some genes are annotated to multiple chromsomes. Here we take the first chromsome each is annotated to. 
```{r}
dim(chr)
dim(edata)
# Take non-duplicated chromsomes
chr = chr[!duplicated(chr[,1]),]

# Confirm that the annotation still is in the right order
all(chr[,1] == rownames(edata))

# Select the chromosome Y samples
edatay = dplyr::filter(edata,chr$CHR=="Y")

# Males have Y chromsome expression as expected
boxplot(colSums(edatay) ~ pdata$gender)
points(colSums(edatay) ~ jitter(as.numeric(pdata$gender)),
        col=as.numeric(pdata$gender),
        pch=19)

```





