---
output:
  BiocStyle::html_document
---


<!-- These options are set up specifically for Bioconductor flavored Markdown --> 


<!-- rmarkdown v1 -->

<!--
%\VignetteIndexEntry{R Markdown Lecture}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}
-->

```{r style, results = 'asis',include=FALSE}
BiocStyle::markdown()
```

```{r global_options,include=FALSE}
## see ch. 10 Hooks of Xie's knitr book
knit_hooks$set(setPch = function(before, options, envir) {
  if(before) par(pch = 19)
})
opts_chunk$set(setPch = TRUE)
library(RSkittleBrewer)
# Make the colors pretty
trop = RSkittleBrewer("tropical")
palette(trop)
```


# Dimension reduction

Package: [genstats](https://github.com/jtleek/genstats)<br />
Author: [Jeff Leek](www.jtleek.com) <br />
Compilation date: `r Sys.Date()`

### General principles

* Can we find patterns in matrices of data?


### Load some data

We will use this expression set that combines two studies [Transcriptome genetics using second generation sequencing in a Caucasian population.](http://www.ncbi.nlm.nih.gov/pubmed?term=20220756%5Buid%5D) and [Understanding mechanisms underlying human gene expression variation with RNA sequencing.](http://www.ncbi.nlm.nih.gov/pubmed?term=20220758). These studies are different populations but we counted the same genes for both. Then we'll explore the differences. 

```{r}
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/montpick_eset.RData")
load(file=con)
close(con)
mp = montpick.eset
pdata=pData(mp)
edata=as.data.frame(exprs(mp))
fdata = fData(mp)
ls()
```

### Calculate the singular vectors

Here we calculate the singular vectors: 

```{r}
edata_centered = edata - rowMeans(edata)
svd1 = svd(edata_centered)
names(svd1)
```

### Look at the percent variance explained

The percent of variance explained is given by $\frac{d_{ii}}{\sum_{j}d_{jj}^2}$
```{r}
plot(svd1$d,ylab="Singular value",col=2)
plot(svd1$d^2/sum(svd1$d^2),ylab="Percent Variance Explained",col=2)
```

### Plot top two principal components

```{r}
par(mfrow=c(1,2))
plot(svd1$v[,1],col=2,ylab="1st PC")
plot(svd1$v[,2],col=2,ylab="2nd PC")
```

### Plot PC1 vs. PC2

A very common plot is to plot PC1 versus PC2 to see if you can see any "clusters" or "groups".

```{r}
plot(svd1$v[,1],svd1$v[,2],col=2,ylab="2nd PC",xlab="1st PC")
```

One thing you can do is color them by different variables to see if clusters stand out. 

```{r}
plot(svd1$v[,1],svd1$v[,2],ylab="2nd PC",
     xlab="1st PC",col=as.numeric(pdata$study))
```

Another common plot is to make boxplots comparing the PC for different levels of known covariates (don't forget to show the actual data!).

```{r}
boxplot(svd1$v[,1] ~ pdata$study,border=c(1,2))
points(svd1$v[,1] ~ jitter(as.numeric(pdata$study)),col=as.numeric(pdata$study))
```


### PCs versus SVs

What we have been plotting is not exactly the principal components. 

```{r}
pc1 = prcomp(edata)
plot(pc1$rotation[,1],svd1$v[,1])
```

To get the actual PCs you have to subtract the column means rather than the row means when normalizing. 

```{r}
edata_centered2 = t(t(edata) - colMeans(edata))
svd2 = svd(edata_centered2)
```

Despite this, it is most common for people to perform row-centering and then plot the singular vectors (sometimes labeling them PCs like I have done in this document)

### Dependencies

These are the packages that this tutorial depends on:

* [BiocStyle](http://www.bioconductor.org/packages/release/bioc/html/BiocStyle.html)
* [knitr](https://cran.r-project.org/package=knitr)
* [RSkittleBrewer](https://github.com/alyssafrazee/RSkittleBrewer)
* [Biobase](http://bioconductor.org/packages/release/bioc/html/Biobase.html)
* [devtools](https://cran.r-project.org/package=devtools)

### Session information

```{r session_info}
devtools::session_info()
```
